### PR TITLE
feat(audio): pull download link from iframe

### LINKED ISSUE #356: [metascraper-audio] Run parser on subframes of page
I've noticed a variety of websites where  the audio URL lives in an iframe and within that iframe the URL follows a format that metascraper understands.

Here are a few examples (link, followed by iframe url where metascraper succeeds & the rule that gets the job done):

https://www.indiehackers.com/podcast/032-peter-and-calvin-of-segment
https://share.transistor.fm/e/70c487ed (`a['href']`)

https://theflip.africa/podcast/digitizing-healthcare-helium-health-goke-olubusi/
https://player.simplecast.com/3ab2ec74-8d89-4344-9be0-a30fa53ac6a7 (`audio['src']`)

https://bfi.uchicago.edu/podcast/the-big-tech-threat/
http://wbez-rss.streamguys1.com/player/player21011316001810372.html (`audio['src']`)

If we were to apply the `metascraper-audio` rules to the contents of the iframe then we could successfully get the audio URLs.

I realize a solution may be relatively complex, so I'm presenting this to you @Kikobeats to see if you have any ideas about how we could achieve this.

Even when I present `metascraper` with an HTML string that contains the the dom of the document inside an `iframe` tag it seems to ignore everything inside of the `iframe` tag.

Seems like this might require running the parser on all frames in the page until something is returned?

### PR DESCRIPTION
**Pull request by goleary**
Closes #356.

Following up on #359.

This one only loads/parses an iframe in the `metascraper-audio` package so as not to affect performance in other parts of `metascraper`.

I copied [this function](https://github.com/microlinkhq/metascraper/blob/a6f6e1dc7a75039f83569e2f531cf8abdd9e0c30/packages/metascraper-telegram/index.js#L26) from `metascraper-telegram` for the actual loading of the iframe since their contents will not be present in the html string provided to metascraper. It should probably be moved to `@metascraper/helpers` if we want to proceed with this approach.

This change only applies the `a['href']` rule (for now).  Instead of duplicating the other audio rules, I'd like to figure out a way to apply all of the audio rules to iframe contents.

[This](https://www.indiehackers.com/podcast/032-peter-and-calvin-of-segment) is the link I've been using to test.

There are still some downside to this approach:
- The iframe contents are loaded within metascraper (using `JSDOM`). If a consumer is using this from a headless browser then the iframe contents may be loaded more than once (once within the headless browser, & another time within `metascraper-audio`)
- This currently only operates on a single `iframe`. Some pages have multiple, so it's possible the first one returned is not the one with the audio link.

### CODE REVIEW

**Comment by Kikobeats** [comment]
Thanks for this! let me spend some time testing the approach and ping you back 🙂

**Comment by Kikobeats** [comment]
I appreciate your effort making `metascraper-audio` even better every day 👏

You took `metascraper-telegram` as inspiration, but keep in mind the logic wait until iframe is load is applied just for telegram URLs that return `true` at test function level:

https://github.com/microlinkhq/metascraper/blob/master/packages/metascraper-telegram/index.js#L80

In this case, since you are not sure if an `iframe` HTML element is present on the markup, you need to check first for that

```js
const hasIframe = $('iframe').length > 0
```

otherwise, you will be waiting forever for an iframe load that never will happen

Also, I'm thinking about how this can be landed. Essentially you want to re-rerun the same rules, but passing the inner HTML from the iframe

File: packages/metascraper-audio/index.js [hunk 1]
```diff
@@ -29,6 +57,14 @@ module.exports = () => ({
     toAudio($jsonld('contentUrl')),
     toAudio($ => $('audio').attr('src')),
     toAudio($ => $('audio > source').attr('src')),
-    ({ htmlDom: $ }) => $filter($, $('a'), el => audio(el.attr('href')))
+    ({ htmlDom: $ }) => $filter($, $('a'), el => audio(el.attr('href'))),
+    toAudio(async ($, url) => {
+      // Duplicated logic to the rule above
+      //TODO: figure out a way to apply ALL audio rules to an iframe instead of
+      // duplicating the rules in an iframe variant
```
Comment (Kikobeats) [inline packages/metascraper-audio/index.js:64]:
```suggestion
      // duplicating the rules in an iframe variant
     if ($('iframe').length === 0) return
```

**Comment by Kikobeats** [comment]
@goleary I didn't test it but this should be works fine to re-use audio rules:

```js
const audioRules = [
  toAudio($ => $('meta[property="og:audio:secure_url"]').attr('content')),
  toAudio($ => $('meta[property="og:audio"]').attr('content')),
  toAudio($ => {
    const contentType = $(
      'meta[property="twitter:player:stream:content_type"]'
    ).attr('content')
    const streamUrl = $('meta[property="twitter:player:stream"]').attr(
      'content'
    )
    return contentType ? withContentType(streamUrl, contentType) : streamUrl
  }),
  toAudio($jsonld('contentUrl')),
  toAudio($ => $('audio').attr('src')),
  toAudio($ => $('audio > source').attr('src')),
  ({ htmlDom: $ }) => $filter($, $('a'), el => audio(el.attr('href')))
]

module.exports = () => ({
  audio: [
    ...audioRules,
    toAudio(async ($, url) => {
      if ($('iframe').length === 0) return

      const dom = await loadIframe(url, $.html())
      const $2 = cheerio.load(dom.document.body.innerHTML)

      let index = 0
      let value

      do {
        const rule = rules[index++]
        value = await rule({ htmlDom: $2, url })
      } while (!has(value) && index < audioRules.length)

      return value
    })
  ]
})
```

**Comment by goleary** [comment]
Awesome, thanks for the feedback & suggestions!

I'll try out something like you suggest above & update this PR.

One question (which I could confirm by looking at code, but I'll ask it anyway):

Once a rule successfully returns audio does that end execution? So all following rules are _not_ run?

Essentially if, `audioRules` (non-iframe) returns audio, then we won't run any of the iframe rules right?

**Comment by Kikobeats** [comment]
the logic will end with the first rule that returns a no falsy value

**Comment by goleary** [comment]
Sweet, that's what I figured. Thanks!

**Comment by goleary** [comment]
I made the changes you recommended @Kikobeats and it works for 2 out of the 3 URLs I've been using to test (you can see the other 2 [here](https://github.com/microlinkhq/metascraper/pull/359)).

There is still no audio returned for this url: https://theflip.africa/podcast/digitizing-healthcare-helium-health-goke-olubusi/

It seems like maybe the iframe scripts are not run because the `<audio>` tag has no `src` when the rule runs. I find this strange given that we create the JSDOM with these options: `{ runScripts: 'dangerously', resources: 'usable' }`, but who knows...

I plan to look into it some more when I have more time.

File: packages/metascraper-audio/index.js [hunk 2]
```diff
@@ -5,30 +5,76 @@ const {
...
+const { JSDOM, VirtualConsole } = require('jsdom')
+const cheerio = require('cheerio')
 
 const toAudio = toRule(audio)
 
 const withContentType = (url, contentType) =>
   isMime(contentType, 'audio') ? url : false
 
+// copied from metascraper-telegram TODO: move to helper
+const loadIframe = asyncMemoizeOne(
```
Comment (Kikobeats) [inline packages/metascraper-audio/index.js:21]: maybe makes sense to move this as part of `@metascraper/helpers` so `metascraper-audio` and `metascraper-telegram` can use it 🙂

**Comment by Kikobeats** [comment]
hmm looks like the current JSDOM configuration is okay

https://github.com/jsdom/jsdom#basic-options

I suspect the website is executing JavaScript for mounting the audio src; in that case, it's a consideration to be taken related to how the HTML is extracted rather than detection.

**Comment by goleary** [comment]
thanks for the refactoring & integration tests @Kikobeats!

I went ahead and marked it ready for merge as any followup from the investigation into https://theflip.africa/podcast/digitizing-healthcare-helium-health-goke-olubusi/ can come in a later PR.

**Comment by Kikobeats** [comment]
@goleary I think the problem is how the audio player is being mounted at `theflip`.

`theflip` uses `simplecast` and I added an integration test for that. It's working fine 🤔

**Comment by Kikobeats** [comment]
@goleary we can try a different approach

1. remove `resources: "usable"` from https://github.com/microlinkhq/metascraper/pull/360/files#diff-5fa62491fdc784b119f866e62f2fa57622727b91416ddaff5e40d65066be95b3R361
2. get the iframe resource by `$('iframe').attr('src')`
3. in case the iframe src URL is present, perform a simple GET for getting the HTML and use it instead.

**Pull request merged**